File: app/src/main/java/com/liadpaz/amp/viewmodels/Album.java
package com.liadpaz.amp.viewmodels;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.NonNull;

import java.util.ArrayList;

public class Album implements Parcelable {
    public static final Creator<Album> CREATOR = new Creator<Album>() {
        @Override
        @NonNull
        public Album createFromParcel(Parcel in) {
            return new Album(in);
        }

        @Override
        @NonNull
        public Album[] newArray(int size) {
            return new Album[size];
        }
    };
    private static final String TAG = "AmpApp.Album";
    public final String name;
    public final String artist;
    public final ArrayList<Song> songs;

    public Album(@NonNull String name, @NonNull String artist, @NonNull ArrayList<Song> songs) {
        this.name = name;
        this.artist = artist;
        this.songs = songs;
    }

    private Album(@NonNull Parcel in) {
        name = in.readString();
        artist = in.readString();
        songs = in.createTypedArrayList(Song.CREATOR);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(@NonNull Parcel dest, int flags) {
        dest.writeString(name);
        dest.writeString(artist);
        dest.writeTypedList(songs);
    }
}

File: app/src/main/java/com/liadpaz/amp/viewmodels/Album.kt
package com.liadpaz.amp.viewmodels

import android.os.Parcel
import android.os.Parcelable
import java.util.*

data class Album(val name: String, val artist: String, val songs: ArrayList<Song>) : Parcelable {

    private constructor(parcel: Parcel) : this(parcel.readString()!!, parcel.readString()!!, parcel.createTypedArrayList(Song.CREATOR)!!)

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(name)
        dest.writeString(artist)
        dest.writeTypedList(songs)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (this.javaClass != other?.javaClass) return false
        other as Album
        return other.artist == artist && other.name == name && other.songs == songs
    }

    override fun hashCode(): Int {
        var result = name.hashCode()
        result = 31 * result + artist.hashCode()
        result = 31 * result + songs.hashCode()
        return result
    }

    companion object CREATOR : Parcelable.Creator<Album> {
        override fun createFromParcel(parcel: Parcel): Album {
            return Album(parcel)
        }

        override fun newArray(size: Int): Array<Album?> {
            return arrayOfNulls(size)
        }
    }
}